cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -43,32 +43,49 @@ bool HashJoinSchema::IsTypeSupported(const DataType& type) {
   return is_fixed_width(id) || is_binary_like(id) || is_large_binary_like(id);
 }
 
-Result<std::vector<FieldRef>> HashJoinSchema::VectorDiff(const Schema& schema,
-                                                         const std::vector<FieldRef>& a,
-                                                         const std::vector<FieldRef>& b) {
-  std::unordered_set<int> b_paths;
-  for (size_t i = 0; i < b.size(); ++i) {
-    ARROW_ASSIGN_OR_RAISE(auto match, b[i].FindOne(schema));
-    b_paths.insert(match[0]);
+Result<std::vector<FieldRef>> HashJoinSchema::ComputePayload(
+    const Schema& schema, const std::vector<FieldRef>& output,
+    const std::vector<FieldRef>& filter, const std::vector<FieldRef>& keys) {
+  // payload = (output + filter) - keys, with no duplicates
```

Comment:
So what happens if there are duplicate field names in the schema?  Is this a malformed query?

@bkietz Is this something the IR would detect and prevent in the first place?

Reply:
IR requires that fields be referenced by index, so it would be up to a producer to decide how to navigate duplicate field names.

Reply:
Ok.  I spoke to Sasha a bit about this as well.  Just to close the loop, in IR, in a join expression, field reference indices are indices into the concatenated {left_schema, right_schema} correct?  So `field_ref(5)` is `left_schema[5]` if `len(left_schema) > 5` and otherwise it is `right_schema[5 % len(left_schema)]`?

Reply:
Should be `right_schema[5 - len(left_schema)]` (the pathological case is if `len(left_schema) == 1`, then using `%` would always evaluate to `0`).

Reply:
Concatenated schema as you describe is what substrait specifies https://substrait.io/relations/logical_relations/#join-operation so we should probably stick with that for simplicity.